### ISSUE TITLE
Question about code in hyper/render.js:53 (upgrade())

### ISSUE THREAD

**Issue by treshugart** [WebReflection/hyperHTML#191]
To give some background, I'm writing a HyperHTML renderer for Skate. All tests tests are run via Jest. However, Jest requires we have a Node DOM implementation, so of course, we wrote one that works with custom elements and shadow DOM: https://github.com/skatejs/skatejs/tree/master/packages/ssr.

When running the tests, I get the following error:

```
TypeError: this[(i - 1)] is not a function
      
  at Array.update (node_modules/hyperhtml/cjs/hyper/render.js:67:16)
  at MyElement.upgrade (node_modules/hyperhtml/cjs/hyper/render.js:58:10)
  at MyElement.render (node_modules/hyperhtml/cjs/hyper/render.js:43:13)
  at MyElement.render (packages/renderer-hyperhtml/src/__tests__/index.js:41:12)
  at MyElement.renderer (packages/renderer-hyperhtml/src/index.js:31:9)
  at Object.<anonymous> (packages/renderer-hyperhtml/src/__tests__/index.js:51:6)
```

The interesting thing is that when I guard the `this[(i - 1)]` invocation to ensure that it's defined prior to doing so, all tests pass just fine. They're just simple integration tests, so I'm not doing anything crazy.

I'm wondering if putting a guard around this would be a good idea? I can't really tell because I'm having a hard time grokking what that code is actually doing and was hoping you could point me in the right direction.

The package is here in case you want a look: https://github.com/skatejs/skatejs/tree/v6/packages/renderer-hyperhtml.

**Comment by WebReflection** [comment]
update is in the critical path, anything in there will impact performance.

this issue looks very related to this one:
https://github.com/WebReflection/hyperHTML/issues/190

mind following those instructions using a node DOM that does support custom elements and also get/setAttributeNode as _basicHTML_ is?

**Comment by WebReflection** [comment]
now that I look at [undom](https://github.com/developit/undom/blob/master/src/undom.js) I can tell you already **many** things are missing. hyperHTML heavily uses DOM standards, including template, cloneNode, comments, etcetera ... I don't think undom is the tool you are looking for, even JSDOM that supposed to be standard has difficulties with all the standard API used in here.

Give [basicHTML](https://github.com/WebReflection/basicHTML) a chance 😉 

Closing here as I believe there's no action for me to take.

**Comment by WebReflection** [comment]
P.S. it'd be awesome if you could contribute to bring ShadowDOM in basicHTML too !!! I always thought that was useless as SSR 'cause ... well, it's shadow.

**Comment by treshugart** [comment]
Yeah, nah.

**Comment by WebReflection** [comment]
> Yeah, nah.

then I don't know how to help. The update expects exact amount of arguments. If that's not preserved, everything else might fail.

TL;DR if you have such error you have either a poor DOM env or you manipulate in an obtrusive way the DOM owned by hyperHTML render/wire logic.

If latter is the case you can have a look at this other ticket:
https://github.com/WebReflection/hyperHTML/issues/166
